incubator/hnc/controllers/hierarchy_controller.go

```diff
@@ -47,76 +49,144 @@ type HierarchyReconciler struct {
+// namespace).
+type NamespaceSyncer interface {
+	SyncNamespace(context.Context, logr.Logger, string) error
 }
 
 // +kubebuilder:rbac:groups=hnc.x-k8s.io,resources=hierarchies,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=hnc.x-k8s.io,resources=hierarchies/status,verbs=get;update;patch
 
-// Reconcile invokes onDelete or update, depending on whether the Hierarchy object exists or not.
+// Reconcile simply calls SyncNamespace, which can also be called if a namespace is created or
+// deleted.
 func (r *HierarchyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
-	log := r.Log.WithValues("trigger", req.NamespacedName.Namespace)
+	nnm := req.NamespacedName
+	log := r.Log.WithValues("trigger", req.NamespacedName)
+	return ctrl.Result{}, r.SyncNamespace(ctx, log, nnm.Namespace)
+}
+
+func (r *HierarchyReconciler) SyncNamespace(ctx context.Context, log logr.Logger, nm string) error {
+	update := true
+	missing := false
 
+	// Get the hierarchy itself.
+	nnm := types.NamespacedName{Namespace: nm, Name: tenancy.Singleton}
 	inst := &tenancy.Hierarchy{}
-	if err := r.Get(ctx, req.NamespacedName, inst); err != nil {
+	if err := r.Get(ctx, nnm, inst); err != nil {
 		if !errors.IsNotFound(err) {
-			log.Info("Couldn't read")
-			return ctrl.Result{}, err
+			log.Error(err, "Couldn't read singleton", "ns", nm)
+			return err
 		}
 
-		return ctrl.Result{}, r.onDelete(ctx, log, req.NamespacedName.Namespace)
+		// It doesn't exist - initialize it to a sane initial value.
+		log.V(1).Info("Missing singleton", "ns", nm)
+		missing = true
+		inst.ObjectMeta.Name = tenancy.Singleton
+		inst.ObjectMeta.Namespace = nm
 	}
 
+	// If the hierarchy object exists but is being deleted, we won't update it when we're finished
+	// syncing (we should sync our internal data structure anyway just in case something's changed).
 	if !inst.GetDeletionTimestamp().IsZero() {
-		// Wait for it to actually be deleted
-		return ctrl.Result{}, nil
+		log.Info("Singleton is being deleted", "ns", nm)
+		update = false
+	}
+
+	// If it's missing, check to see if the namespace is missing as well. If so, we need to remove
+	// this namespace from the forest.
+	if missing {
+		ns := &corev1.Namespace{}
+		nsnnm := types.NamespacedName{Name: nm}
+		if err := r.Get(ctx, nsnnm, ns); err != nil {
+			if !errors.IsNotFound(err) {
+				log.Error(err, "Couldn't read namespace", "ns", nm)
+				return err
+			}
+
+			// The namespace is gone. Remove all traces of the namespace from  our data structures.
+			log.Info("Namespace was deleted", "ns", nm)
+			return r.onDelete(ctx, log, nm)
+		}
+
+		// If the namespace is being deleted, we should still sync any changes that we might have missed
+		// (ie, maybe the parent used to be set?) but then don't update anything.
+		if !ns.GetDeletionTimestamp().IsZero() {
+			log.Info("Namespace is being deleted", "ns", nm)
+			update = false
```

Comment:
Why do we need this check? If ns is being deleted. A reconcile will eventually happen and go to r.onDelete path to clear everything.

Reply:
TBH I wasn't sure what to do with this. I think this check showed up in the kubebuilder boilerplate code and I just filled it in with something that seemed reasonable. I should probably update the comments to indicate that I don't really know what I'm doing.

Reply:
Are you going to change this later?

Reply:
I'll update the comments in another commit, but I'll leave this as-is unless someone has a better idea.

While I understand that a reconcile will _eventually_ happen, my idea here was this could take a while (eg namespaces take a while to be deleted) and we should update everything else ASAP.